cpp/src/decisiontree/batched-levelalgo/kernels/builder_kernels.cuh

```diff
@@ -111,6 +122,209 @@ HDI IdxT lower_bound(DataT* array, IdxT len, DataT element)
+  for (int i = 0; i < MAX_SAMPLES_PER_THREAD; ++i)
+    mask[i] = 0;
+
+  do {
+    // blocked arrangement
+    for (int cta_sample_idx = MAX_SAMPLES_PER_THREAD * threadIdx.x, thread_local_sample_idx = 0;
+         thread_local_sample_idx < MAX_SAMPLES_PER_THREAD;
+         ++cta_sample_idx, ++thread_local_sample_idx) {
+      // mask of the previous iteration, if exists, is re-used here
+      // so previously generated unique random numbers are used.
+      // newly generated random numbers may or may not duplicate the previously generated ones
+      // but this ensures some forward progress in order to generate atleast 'k' unique random
+      // samples.
+      if (mask[thread_local_sample_idx] == 0 and cta_sample_idx < n_parallel_samples)
+        raft::random::custom_next(
+          gen, &items[thread_local_sample_idx], uniform_int_dist_params, IdxT(0), IdxT(0));
+      else if (mask[thread_local_sample_idx] ==
+               0)  // indices that exceed `n_parallel_samples` will not generate
+        items[thread_local_sample_idx] = n - 1;
+      else
+        continue;  // this case is for samples whose mask == 1 (saving previous iteraion's random
+                   // number generated)
+    }
+
+    // Specialize BlockRadixSort type for our thread block
+    typedef cub::BlockRadixSort<IdxT, BLOCK_THREADS, MAX_SAMPLES_PER_THREAD> BlockRadixSortT;
+    // BlockAdjacentDifference
+    typedef cub::BlockAdjacentDifference<IdxT, BLOCK_THREADS> BlockAdjacentDifferenceT;
+    // BlockScan
+    typedef cub::BlockScan<IdxT, BLOCK_THREADS> BlockScanT;
+
+    // Shared memory
+    __shared__ union TempStorage {
+      typename BlockRadixSortT::TempStorage sort;
+      typename BlockAdjacentDifferenceT::TempStorage diff;
+      typename BlockScanT::TempStorage scan;
+    } temp_storage;
+
+    // collectively sort items
+    BlockRadixSortT(temp_storage.sort).Sort(items);
+
+    __syncthreads();
+
+    // compute the mask
+    // compute the adjacent differences according to the functor
+    BlockAdjacentDifferenceT(temp_storage.diff)
+      .FlagHeads(mask, items, mask, CustomDifference<IdxT>());
+
+    __syncthreads();
+
+    // do a scan on the mask to get the indices for gathering
+    BlockScanT(temp_storage.scan).ExclusiveSum(mask, col_indices, n_uniques);
+
+    __syncthreads();
+
+  } while (n_uniques < k);
+
+  // write the items[] of only the ones with mask[]=1 to col[offset + col_idx[]]
+  IdxT col_offset = k * blockIdx.x;
+  for (int i = 0; i < MAX_SAMPLES_PER_THREAD; ++i) {
+    if (mask[i] and col_indices[i] < k) { colids[col_offset + col_indices[i]] = items[i]; }
+  }
+}
+
+// algo L of the reservoir sampling algorithm
+/**
+ * @brief Generates 'k' unique samples of features from 'n' feature sample-space using the algo-L
+ * algorithm of reservoir sampling. wiki :
+ * https://en.wikipedia.org/wiki/Reservoir_sampling#An_optimal_algorithm
+ */
+template <typename IdxT>
+__global__ void algo_L_sample_kernel(int* colids,
```

Comment:
Ideally this (and possibly other sampling methods here) should be generalized to become the batched counterpart of [sampleWithoutReplacement](https://github.com/rapidsai/raft/blob/4f22957db9a38d81d0cd6ad864507c47d243ad56/cpp/include/raft/random/rng.cuh#L705). What about creating an issue for this?

Reply:
[Done](https://github.com/rapidsai/raft/issues/767).